### PR TITLE
[Zhao Peiduo] iP

### PR DESCRIPTION
**Pull request by ZhaoPeiduo**
# Implementation for Duke

> This document presents the avaliable features and user commands. The following information may be used as a user guide for my implementaton of Duke. The original version of Duke is avaliable [here](https://github.com/nus-cs2103-AY2122S1/ip).

### Features :ballot_box_with_check:

- [x] Save and delete different types of tasks: todo, deadline, event;
- [x] Set a saved task to be done;
- [x] Autosave to and autoload user's data from a default directory;
- [x] Allows user to save/load data from a specified directory;
- [x] Search for task according to a specified key word.
- [x]  An in-app user guideline 

- todo <**description**>: create and save a todo task;
- deadline <**description**> /by <**time in format yyyy-mm-dd**>: create and save a deadline task;
- event <**description**> /at <**time in format yyyy-mm-dd**>: create and save an event task;
- save <**directory**>: save the current task list to the specified directory;
- load <**directory**>: load the task list from the specified directory;
- done <*number*>: mark the task with *number* index as done;
- delete <*number*>: delete the task with *number* index;
- deleteAll: delete all saved tasks in the task list;
- find <**keyword**>: find all tasks with description containing the **keyword**.
- help: shows the list of avaliable command

You may run the application by clicking the green triangle on the left of the main function in `Duke.class`.
```
public class Duke {
    ...
    public static void main(String[] args) {
        new Duke().run();
    }
}
```

### CODE REVIEW

File: src/main/java/duke/Deadline.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
+package duke;
+
+import java.time.LocalDate;
+
+/**
+ * Represents a specific type of Task with an additional deadline(ddl) field.
+ */
+public class Deadline extends Task {
+    private LocalDate ddl;
```
Comment (marcuspeh) [inline src/main/java/duke/Deadline.java]: Is there any reason why this variable is ddl? Perhaps it will be better if it can be changed to a name that explains the name entiry.

File: src/main/java/duke/Parser.java [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+            builder.append("deleteAll\n");
+            builder.append("find <keyword>\n");
+            return builder.toString();
+        } else if (userInput.startsWith("todo ") || userInput.startsWith("deadline ")
+                || userInput.startsWith("event ")) {
+            return tasks.add(userInput, tasks.getStorage().getUserInputRecords());
+        } else {
+            return "OOPS!!! I'm sorry, but I don't know what that means :-(\n";
+        }
+    }
```
Comment (marcuspeh) [inline src/main/java/duke/Parser.java]: Maybe you can abstract this method more? 
image

File: src/main/java/duke/Storage.java [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                        task.setDone(true);
+                    }
+                    userInputRecords.add(task);
+                }
+            }
+            scanner.close();
+        } catch (IOException e) {
+            System.out.println("Saved data not found, a new data file created.\n");
+        }
+    }
```
Comment (marcuspeh) [inline src/main/java/duke/Storage.java]: Perhaps you can abstract or simplify this method more. It's quite hard to follow the logic behind this with so much with the nested while and if(s) loops.

File: src/main/java/duke/Deadline.java [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+     * Two deadline objects are equal iff they are equal tasks and they have the same ddl.
+     * */
+    @Override
+    public boolean equals(Object another) {
+        if (another instanceof Deadline) {
+            Deadline anotherTask = (Deadline) another;
+            return super.equals(another) && ddl.equals(anotherTask.getDdl());
+        }
+        return false;
+    }
```
Comment (marcuspeh) [inline src/main/java/duke/Deadline.java]: Was this equals method used anywhere?

File: src/main/java/duke/Event.java [hunk 5]
```diff
@@ -0,0 +1,45 @@
+package duke;
+
+import java.time.LocalDate;
+
+/**
+ * Represents a specific type of Task with an additional time field.
+ */
+public class Event extends Task {
+    private LocalDate time;
```
Comment (marcuspeh) [inline src/main/java/duke/Event.java]: Since this is saving the date instead of time, maybe you can rename this variable to date instead.

**Review by marcuspeh — COMMENTED** [review]
Overall quite well done. I can hardly find anything to comment on and those that I mentioned is more of trivial improvements. Good job!

File: src/main/java/duke/Deadline.java (hunk 1 above)
Comment (ZhaoPeiduo) [reply src/main/java/duke/Deadline.java]: Thanks for pointing this out. I have changed the variable name to deadline in the latest commit!

File: src/main/java/duke/Parser.java (hunk 2 above)
Comment (ZhaoPeiduo) [reply src/main/java/duke/Parser.java]: Thanks for pointing this out. I have abstracted the construction of help message to anothe private method in the latest commit!

File: src/main/java/duke/Deadline.java (hunk 4 above)
Comment (ZhaoPeiduo) [reply src/main/java/duke/Deadline.java]: Yes, it is used in the JUnit test ParserTasklistTest to compare whether the expected and actual list of tasks are equal.

File: src/main/java/duke/Event.java (hunk 5 above)
Comment (ZhaoPeiduo) [reply src/main/java/duke/Event.java]: Thanks for pointing this out. I have changed the variable name to date in the latest commit!

**Comment by ZhaoPeiduo** [comment]
@marcuspeh Thanks for the helpful comments!

File: src/main/java/duke/Storage.java (hunk 3 above)
Comment (ZhaoPeiduo) [reply src/main/java/duke/Storage.java]: Thanks for the suggestion! I have abstracted out the add task process to another private method in the latest commit!

File: src/main/java/DialogBox.java [hunk 6]
```diff
@@ -0,0 +1,72 @@
+import java.awt.*;
```
Comment (yunpeng1234) [inline src/main/java/DialogBox.java]: Should not import all as * as stated in the coding violation. Also, all classes in should in a package so the package statement is missing here.

File: src/main/java/DialogBox.java [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+import javafx.collections.ObservableList;
+import javafx.fxml.FXML;
+import javafx.fxml.FXMLLoader;
+import javafx.geometry.Insets;
+import javafx.geometry.Pos;
+import javafx.scene.Node;
+import javafx.scene.control.Label;
+import javafx.scene.image.Image;
+import javafx.scene.image.ImageView;
+import javafx.scene.layout.*;
```
Comment (yunpeng1234) [inline src/main/java/DialogBox.java]: Likewise, here the use of * is not allowed.

File: src/main/java/DialogBox.java [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+import javafx.scene.paint.Paint;
+import javafx.scene.shape.Circle;
+
+// This class is taken from JavaFX tutorial 4 by Jeffry Lum https://se-education.org/guides/tutorials/javaFxPart4.html
```
Comment (yunpeng1234) [inline src/main/java/DialogBox.java]: Redundant comment that should be deleted.

File: src/main/java/MainWindow.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
+import duke.Duke;
+import javafx.fxml.FXML;
+import javafx.geometry.Insets;
+import javafx.scene.control.Button;
+import javafx.scene.control.ScrollPane;
+import javafx.scene.control.TextField;
+import javafx.scene.image.Image;
+import javafx.scene.layout.*;
```
Comment (yunpeng1234) [inline src/main/java/MainWindow.java]: Same issue as above.

File: src/main/java/duke/Parser.java [hunk 10]
```diff
@@ -0,0 +1,129 @@
...
+    /**
+     * This method checks if a String starting with delete is a delete command.
+     * A delete command is a string with the keyword delete followed by numbers and spaces only.
+     * Examples of valid user inputs:
+     * -> delete 3
+     * -> delete   4(spaces trimmed)
+     * -> delete 1 4(spaces trimmed and interpret as done 14)
+     * Examples of invalid user inputs:
+     * -> delete task
+     * -> delete 2 task
```
Comment (yunpeng1234) [inline src/main/java/duke/Parser.java]: Description of the method should be simple. Additional comments on the code can be done in the readme.md

File: src/main/java/duke/TaskList.java [hunk 11]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * Marks the user-specified event as done, according to the task index.
+     *
+     * @param userInput        input from the user.
+     * @param userInputRecords the saved record.
+     * @return the response on whether a task is successfully marked done.
+     */
+    public String markAsDone(String userInput, ArrayList<Task> userInputRecords) {
+        try {
+            int itemToComplete = Integer.parseInt(userInput.replaceAll("[^0-9]", "")) - 1;
```
Comment (yunpeng1234) [inline src/main/java/duke/TaskList.java]: Variable could be renamed to sound like a number. Eg. ItemIndex

File: src/main/java/duke/TaskList.java [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+    /**
+     * Searches for the events according to the user-specified keyword.
+     *
+     * @param userInput        input from the user.
+     * @param userInputRecords the saved record.
+     * @return the response on whether a task is successfully found.
+     */
+    public String search(String userInput, ArrayList<Task> userInputRecords) {
+        String keyword = userInput.replace("find", "").trim();
+        ArrayList<Task> searchResult = new ArrayList<>();
```
Comment (yunpeng1234) [inline src/main/java/duke/TaskList.java]: Could pluralise the search result for >1 returned result

**Review by yunpeng1234 — COMMENTED** [review]
Overall no major coding standard violation. LGTM.

Quality wise, the code is bloated with own LocalDate formatter, I suggest you use their formatter to cut down on code bloat. 
Testing of storage should be simple, no try catch while loops inside. Suggest to use stub to extract of storage for easier testing as well as more readable tests.

**Comment by ZhaoPeiduo** [comment]
@yunpeng1234 Thanks for the helpful suggestions! I have improved my code based on your suggestions
